### DISCUSSION TITLE
Help in understanding the mido messages

### DISCUSSION THREAD

**Discussion by sharmilamani** [mido/mido#585] [General]
Hi, 
How to identify the start time and duration of note played from mid file ?
as i went through the mido read me time in message refers to wait time from previous event. is this the start time of the note ? also i don't see a "note_off message". 

Below is a midi file printed using mido. In this file, the key c4 is pressed and released 10 times.
Kindly help in understanding 
                          1) start of each key press in seconds ?
                          2) time duration / end of each key press ?
                          3) time gap between two key press ( start of second key - end of first key)

```py
MidiFile(type=1, ticks_per_beat=24, tracks=[
  MidiTrack([
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=24, notated_32nd_notes_per_beat=8, time=0),
    MetaMessage('key_signature', key='C', time=0),
    MetaMessage('end_of_track', time=0)]),
  MidiTrack([
    Message('program_change', channel=0, program=4, time=0),
    Message('note_on', channel=0, note=60, velocity=46, time=54),
    Message('note_on', channel=0, note=60, velocity=0, time=30),
    Message('note_on', channel=0, note=60, velocity=40, time=30),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=33, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=38, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=34, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=40, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=45, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=48, time=48),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=51, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=28, time=24),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    MetaMessage('end_of_track', time=0)])
```

**Comment by JorisMonnet** [comment]
Hello,

1.  time in second = tick * tempo * 1e-6  / tick_per_beat
In your case: time = tick * 500000 * 1e-6 / 480
you can find the function tick2second [here](https://github.com/mido/mido/blob/main/mido/midifiles/units.py) for this conversion.

a) note_on messages with velocity=0 are note_off messages. This is a feature of MIDI specification, not a decision from mido.

2. First note -> 30 ticks, then 36,36,36,42...

3. it is the time attribute of the next note_on message -> 54,30,36,36....
